File: cpp-alg/leecode/Q141_linked_list_cycle.cpp
#include <iostream>
#include <unordered_set>

struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(nullptr) {}
};

bool hasCycle(ListNode *head) {
  std::unordered_set<ListNode *> set;
  while (head) {
    if (set.count(head))
      return true;
    set.insert(head);
    head = head->next;
  }
  return false;
}

int main() {

  ListNode *nodeD = new ListNode(4);
  ListNode *nodeC = new ListNode(3);
  ListNode *nodeB = new ListNode(2);
  ListNode *nodeA = new ListNode(1);
  nodeA->next = nodeB;
  nodeB->next = nodeC;
  nodeC->next = nodeD;
  nodeD->next = nullptr;

  bool result = hasCycle(nodeA);
  std::cout << result << std::endl;

  return 0;
}

File: python-alg/leecode/q141_linked_list_cycle.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


def linked_list_cycle(node):
    node_set = set()
    while node:
        if node in node_set:
            return True
        node_set.add(node)
        node = node.next
    return False


nodeD = ListNode(4)
nodeD.next = nodeD
nodeC = ListNode(3)
nodeC.next = nodeD
nodeB = ListNode(2)
nodeB.next = nodeC
nodeA = ListNode(1)
nodeA.next = nodeB

print(linked_list_cycle(nodeA))
